### PR TITLE
Add Semi-SL MeanTeacher for semantic segmentation

### PR DESCRIPTION
**Pull request by kprokofi**
### Summary

This PR intends to add Semi-SL functionality for semantic segmentation task. This PR doesn't consider refactoring changes.

- [x] Add MeanTeacher algorithm
- [x] Semi-SL works for semantic segmentation
- [x] Conduct experiments. Results are better then supervised training at least on few percent
- [x] Add unit tests, integration and performance

In the next PRs the following changes are expected:

- Semantic Segmentation refactoring including aligning models with TrainType introduced by @harimkang. This PR doesn't include this change and simply add new segmentation model with semi_sl suffix
- Prototype based approach as in OTX1.6. Additional experiments are required

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/callbacks/ema_mean_teacher.py
Comment (harimkang) [inline src/otx/algo/callbacks/ema_mean_teacher.py:1]: I think we need a unit test for this file.

File: src/otx/algo/segmentation/dino_v2_seg.py [hunk 1]
```diff
@@ -68,3 +70,46 @@ def to(self, *args, **kwargs) -> Self:
...
+        decode_head_configuration = DinoV2Seg.default_decode_head_configuration | self.decode_head_configuration
+        backbone = DinoVisionTransformer(**backbone_configuration)
+        decode_head = FCNHead(num_classes=self.num_classes, **decode_head_configuration)
+        base_model = DinoV2Seg(
+            backbone=backbone,
+            decode_head=decode_head,
+            criterion_configuration=self.criterion_configuration,
+        )
+
+        return MeanTeacher(base_model, unsup_weight=0.7, drop_unrel_pixels_percent=20, semisl_start_epoch=2)
```
Comment (harimkang) [inline src/otx/algo/segmentation/dino_v2_seg.py:115]: `unsup_weight`, `drop_unrel_pixels_percent`, `semisl_start_epoch` - Do we need to make these values configurable?

File: src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+data_format: common_semantic_segmentation_with_subset_dirs
+include_polygons: true
+unannotated_items_ratio: 0.0
+ignore_index: 255
+train_subset:
+  subset_name: train
+  batch_size: 8
+  num_workers: 4
+  transform_lib_type: TORCHVISION
+  to_tv_image: true
```
Comment (harimkang) [inline src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml]: As far as I know, the Transform that exists in otx can take input as a numpy array, so we don't need to replace it with `tv_image`. So `False` seems appropriate for this, could you please confirm?

As it is, expect one more unnecessary transformation to happen: numpy array -> tv_image -> numpy array.

File: src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        std: [58.395, 57.12, 57.375]
+  sampler:
+    class_path: torch.utils.data.RandomSampler
+
+val_subset:
+  subset_name: val
+  batch_size: 8
+  num_workers: 4
+  transform_lib_type: TORCHVISION
+  to_tv_image: true
```
Comment (harimkang) [inline src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml]: Same here

File: src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        std: [58.395, 57.12, 57.375]
+  sampler:
+    class_path: torch.utils.data.RandomSampler
+
+test_subset:
+  subset_name: test
+  num_workers: 4
+  batch_size: 8
+  transform_lib_type: TORCHVISION
+  to_tv_image: true
```
Comment (harimkang) [inline src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml]: Same here

**Review by harimkang — COMMENTED** [review]
Overall, it looks good, and I have a few comments.

File: src/otx/algo/segmentation/dino_v2_seg.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/dino_v2_seg.py:115]: Yeah, I thought about that, but didn't come up the final solution with that. Could we change this in my following PR with Segmentation task refactoring?

File: src/otx/algo/callbacks/ema_mean_teacher.py
Comment (kprokofi) [reply src/otx/algo/callbacks/ema_mean_teacher.py:1]: Sure, I will add it

File: src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml (hunk 3 above)
Comment (kprokofi) [reply src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml]: Done https://github.com/openvinotoolkit/training_extensions/pull/3801/commits/fc917c8d055333da9ad03852fbdc686e92ef896d

File: src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml (hunk 2 above)
Comment (kprokofi) [reply src/otx/recipe/_base_/data/semisl/semantic_segmentation_semisl.yaml]: https://github.com/openvinotoolkit/training_extensions/pull/3801/commits/fc917c8d055333da9ad03852fbdc686e92ef896d

File: src/otx/algo/callbacks/ema_mean_teacher.py
Comment (kprokofi) [reply src/otx/algo/callbacks/ema_mean_teacher.py:1]: Added https://github.com/openvinotoolkit/training_extensions/pull/3801/commits/5b7cb8a5529fc8a4495dfbd9c03bbc5c58c6bab5

**Review by harimkang — DISMISSED** [review]

**Comment by kprokofi** [comment]
Pascal-VOC results:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


  | Supervised | Semi-Supervised
-- | -- | --
all | 0.6595 | 0.7192
dino_v2 | 0.8634 | 0.8768
litehrnet_18 | 0.5025 | 0.5978
litehrnet_s | 0.4241 | 0.5376
litehrnet_x | 0.3675 | 0.5185
segnext_b | 0.8475 | 0.8520
segnext_s | 0.8171 | 0.8409
segnext_t | 0.7942 | 0.8109



</body>

</html>

DIS5K + Cityscapes results
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


  | Supervised | Semi-Supervised
-- | -- | --
all | 0.7213 | 0.7327
dino_v2 | 0.7875 | 0.7979
litehrnet_18 | 0.6627 | 0.6802
litehrnet_s | 0.6389 | 0.6605
litehrnet_x | 0.6930 | 0.7107
segnext_b | 0.7759 | 0.7627
segnext_s | 0.7571 | 0.7825
segnext_t | 0.7337 | 0.7345



</body>

</html>

The results were averaged across 5 runs with different seeds (OTX perf tests)

Unfortunately, there is no always good result. For example, when labeled images are only 6-20, the pseudo-labels are not reliable and we can observe degradation in metric. Some other approaches / additional fine-tuning is required in that case:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


  | Supervised | Semi-Supervised
-- | -- | --
all | 0.7693 | 0.7267
dino_v2 | 0.8560 | 0.8319
litehrnet_18 | 0.7548 | 0.7540
litehrnet_s | 0.7261 | 0.6988
litehrnet_x | 0.7378 | 0.7343
segnext_b | 0.7805 | 0.6942
segnext_s | 0.7718 | 0.7293
segnext_t | 0.7584 | 0.6445



</body>

</html>

**Review by harimkang — DISMISSED** [review]

File: src/otx/algo/common/utils/utils.py [hunk 5]
```diff
@@ -259,3 +260,69 @@ def inverse_sigmoid(x: Tensor, eps: float = 1e-5) -> Tensor:
     x1 = x.clamp(min=eps)
     x2 = (1 - x).clamp(min=eps)
     return torch.log(x1 / x2)
+
+
+def cut_mixer(images: Tensor, masks: Tensor) -> tuple[Tensor, Tensor]:
```
Comment (sungchul1) [inline src/otx/algo/common/utils/utils.py:265]: It seems to be used just for semantic segmentation, how about moving into semantic segmentation?

File: src/otx/algo/callbacks/ema_mean_teacher.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
+"""Module for exponential moving average for SemiSL mean teacher algorithm."""
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
```
Comment (sungchul1) [inline src/otx/algo/callbacks/ema_mean_teacher.py]:
```suggestion
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
#
"""Module for exponential moving average for SemiSL mean teacher algorithm."""
```

File: src/otx/algo/segmentation/segmentors/mean_teacher.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
+# Copyright (C) 2023 Intel Corporation
```
Comment (sungchul1) [inline src/otx/algo/segmentation/segmentors/mean_teacher.py]:
```suggestion
# Copyright (C) 2024 Intel Corporation
```

File: src/otx/algo/segmentation/segmentors/base_model.py [hunk 8]
```diff
@@ -67,45 +67,69 @@ def forward(
...
+        img_metas: list[ImageInfo],
+        masks: Tensor,
+        interpolate: bool,
+    ) -> Tensor:
+        """Calculates the loss of the model.
+
+        Args:
+            model_features: model outputs of the model.
+            img_metas: Image meta information. Defaults to None.
+            masks: Ground truth masks for training. Defaults to None.
```
Comment (sungchul1) [inline src/otx/algo/segmentation/segmentors/base_model.py]: Please add type annotations.

File: src/otx/algo/common/utils/utils.py (hunk 5 above)
Comment (kprokofi) [reply src/otx/algo/common/utils/utils.py:265]: It will be actually used for other tasks with Semi-SL based on the MeanTeacher approach that we will add in the future. (Object Detection and Instance Segmentation). That is why I added this function to common

File: src/otx/algo/segmentation/segmentors/mean_teacher.py (hunk 7 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/segmentors/mean_teacher.py]: Done https://github.com/openvinotoolkit/training_extensions/pull/3801/commits/c267bbf9e9687b47ee26d92ad855558475301d45

File: src/otx/algo/segmentation/segmentors/base_model.py (hunk 8 above)
Comment (kprokofi) [reply src/otx/algo/segmentation/segmentors/base_model.py]: Done https://github.com/openvinotoolkit/training_extensions/pull/3801/commits/1be0a39d7aea38adb077c8f3d8ad221d9d947dd4

File: src/otx/algo/callbacks/ema_mean_teacher.py (hunk 6 above)
Comment (kprokofi) [reply src/otx/algo/callbacks/ema_mean_teacher.py]: Done https://github.com/openvinotoolkit/training_extensions/pull/3801/commits/1be0a39d7aea38adb077c8f3d8ad221d9d947dd4

File: src/otx/algo/common/utils/utils.py (hunk 5 above)
Comment (kprokofi) [reply src/otx/algo/common/utils/utils.py:265]: However, you are right that it is related to segmentation mostly due to operations with masks. Since I started refactoring task, can I move it in that PR? I would like to merge this PR since my changes depends on it

**Review by harimkang — DISMISSED** [review]

File: src/otx/recipe/_base_/data/semisl/torchvision_semisl.yaml
Comment (sungchul1) [inline src/otx/recipe/_base_/data/semisl/torchvision_semisl.yaml:1]: What do you think about removing `_semisl` suffix for all files in `semisl` directories?
I think we can distinguish they are about semisl with the directory name.

Comment (kprokofi) [reply src/otx/recipe/_base_/data/semisl/torchvision_semisl.yaml:1]: Yeah, maybe we can do that. Let me check it and update with next PR also. 
My main concern is that the model names overlapping during "otx find" command.
image

Comment (harimkang) [reply src/otx/recipe/_base_/data/semisl/torchvision_semisl.yaml:1]: Yeah, we'll have to think about that.. :)

**Review by sungchul1 — DISMISSED** [review]

**Review by sovrasov — APPROVED** [review]

**Review by negvet — APPROVED** [review]

**Review by GalyaZalesskaya — APPROVED** [review]
